### PR TITLE
Update to Lucene 5.4 GeoPointField Type

### PR DESCRIPTION
**Pull request by nknize**
This PR updates ES 3.0 geo_point field types to use Lucene GeoPointField instead of legacy String and parallel DoubleFields. The following is affected:
- GeoPointFieldMapper - updated to use Lucene 5.4 GeoPointField as default indexing type. lat_lon and geohash are preserved as additional indexing options.
- GeoPointFieldData and GeoPointDV - updated to use Lucene 5.4 GeoPointField instead of 2 parallel Double Arrays.
- GeoQueryBuilder/Parser - updated to use Lucene 5.4 GeoPoint\* Queries.
- Unit and Integration Tests - updated to test based on new indexing format.
#### Notes and Todo:
- Need to revisit/update java docs to remove deprecated query options (e.g., optimize_bbox, coerce)
- This PR is for 3.0 only. It is the same as the split 2.x PRs with the exception that it completely removes backward compatibility with 1.x - 2.1 `geo_point` indexes.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/mapper/geo/GeoPointFieldMapper.java [hunk 1]
```diff
@@ -251,6 +236,12 @@ public GeoPointFieldMapper build(BuilderContext context) {
                 } else if (propName.equals("lat_lon")) {
                     builder.enableLatLon(XContentMapValues.nodeBooleanValue(propNode));
                     iterator.remove();
+                } else if (propName.equals("precision_step")) {
+                    builder.precisionStep(XContentMapValues.nodeIntegerValue(propNode));
+                    iterator.remove();
+                } else if (propName.equals("path") && parserContext.indexVersionCreated().before(Version.V_2_0_0_beta1)) {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/mapper/geo/GeoPointFieldMapper.java]: I think this can be removed since this is 3.0?

**Comment by rjernst** [comment]
I quickly glanced at this. It seems like this completely breaks backcompat for geo fields created in 2.x? I think we need to maintain bwc for 2.x in 3.x?

**Comment by nknize** [comment]
> This PR is for 3.0 only. It is the same as the split 2.x PRs with the exception that it completely removes backward compatibility with 1.x - 2.1 geo_point indexes.

It does break bwc per the PR note above. I had an open question/discussion re: geo_point v2 and how long to carry the deprecated ```geo_point``` functionality. Since geopoint v2 slid from the 2.0 release to 2.2 we're required by convention to support bwc until 4.0, no? That seems like a long time, and removing in 3.0 per the original plan still seemed reasonable but I'm OK carrying bwc if we feel its the right thing to do. /cc @clintongormley

**Comment by nknize** [comment]
@s1monw For now it may be easier to add GeoPointV2 (backcompat and all) to both master and 2.x, and open a separate issue to discuss backcompat removal in 3.0? /cc @clintongormley

**Comment by nknize** [comment]
Closing this for now. We can defer the if/when to remove backcompat decision at a later time. /cc @clintongormley

**Pull request closed without merging**